<a href="https://colab.research.google.com/github/linatucker/DISSERTATION/blob/main/Olufemi_and_Lina%2C_TeamFL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# GROUP MEMBERS: 
**OLUFEMI ALLEN-TAYLOR (2008837)**

**LINA PEPPLE (2013592)**


The codes below is to install the required packages used.

In [ ]:
pip install bar_chart_race

     |████████████████████████████████| 163kB 3.1MB/s 


In [ ]:
pip install pycountry

     |████████████████████████████████| 10.1MB 3.0MB/s 
  Created wheel for pycountry: filename=pycountry-20.7.3-py2.py3-none-any.whl size=10746883 sha256=3a2902853dffd7060a040a4a8ef17137ba96b35c61a3aed37b29c0a1487b6b5d
  Stored in directory: /root/.cache/pip/wheels/33/4e/a6/be297e6b83567e537bed9df4a93f8590ec01c1acfbcd405348
Successfully built pycountry


In [ ]:
pip install --upgrade plotly

     |████████████████████████████████| 13.2MB 322kB/s 
  Found existing installation: plotly 4.4.1
    Uninstalling plotly-4.4.1:
      Successfully uninstalled plotly-4.4.1


In [ ]:
# here we import all the required libraries
import pandas as pd
import numpy as np
import pycountry
import plotly.express as px
import matplotlib
import plotly.graph_objects as go
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
import matplotlib.animation as animation
from IPython.display import HTML
import bar_chart_race as bcr

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving mapbox.txt to mapbox.txt


In [ ]:
# Importing the token
px.set_mapbox_access_token(open("mapbox.txt").read())

# DATA SELECTION

**BUSINESS CASE**

Since the end of 2019, the novel COVID-19 virus has been a major threat to the world as the pandemic spread across the globe. Eversince the first case was reported in Wuhan, China, the number of cases and deaths have been on the rise. As a result of the outbreak, several meaasures were implemented to help contain the spread, including ongoing testing, precautionary measures such as sanitation of the environment, lockdown, handwasing, and most recently vaccination.

For our group, it is important to explore the progress of COVID-19 by analysing the data collected from the Ourworldindata and the United Nations, in order to use the top countries with the highest GDP to explore and observe this trend.

The aim of our analysis  is to show the impact of a country's Health Expenditure on the COVID-19 death per population as well as the impact of GDP on vaccination so far, as we will like to see if countries with higher GDP will have a higher % of their population already vaccinated or not. Furthermore, we will see if the Male Life Expectancy of a population has any influence on the % of lives lost per population of a country during the pandemic. Also, according to The Centre For Disease Control [cdc.gov](https://www.cdc.gov/coronavirus/2019-ncov/need-extra-precautions/older-adults.html), older people are more likely to lose thier lives to the COVID-19 virus than the younger folks. Therefore, it is paramount to see if COVID-19 deaths are higher in countries with a large share of older people. Finally, due to the high number of Covid cases, it is important to examine the impact of hospital beds availability on the % of deaths per population. 

**SELECTED DATA**

We used 3 datasets for this project.

Firstly,

The United Nations dataset which was gotten downloaded from [Kaggle](https://www.kaggle.com/sudalairajkumar/undata-country-profiles) and it was originally uploaded on [UNdata](http://data.un.org/). Out of the 50 columns in the data, we will only use 4 columns which are

1.   Country - Countries of the world
2.   GDP: Gross domestic product (million current US$)
3.   Life expectancy at birth (females/males, years) 
4.   Health: Total expenditure (% of GDP)
  
  We have chosen this dataset because it contains all of the parameters that are import for our analysis and it also has all the entries of the world.

Secondly,

The "Our World In Data" dataset. This Data set was originally posted on the "our world in data" website [ourwordindata.org](https://ourworldindata.org/coronavirus-source-data) and futher downloaded and saved in our github repository via the same github [github.com](https://github.com/owid/covid-19-data/tree/master/public/data) posted on the website page and it contains information about Covid-19 compiled from different sources including the World Health Organization since the begining of the pandemic. This dataset gets updated everyday on the ourworld in data website but ours will include entries until the 25th of MARCH 2020 Here is also a link to the description of each column [github](https://github.com/owid/covid-19-data/blob/master/public/data/owid-covid-codebook.csv). It is a very large data set but we will be using just 11 columns for our analysis. These columns are as follows:

1.   'Date' - Date of observation
2.   'iso_code' - ISO 3166-1 alpha-3 – three-letter country codes
3.   'coninent' - Continent of Geographical location
4.   'location' - Geographical location
5.   'new_cases' - New confirmed cases of COVID-19
6.   'new_deaths' - New deaths attributed to COVID-19
7.   'new_tests' - New tests for COVID-19 (only calculated for consecutive days)
8.   'new_vaccinations' - New COVID-19 vaccination doses administered (only calculated for consecutive days)
9.   'hospital_beds_per_thousand' - Hospital beds per 1,000 people, most recent year available since 2010
10.  'population' - Population of each countries
11.  'aged_65_older' - Share of the population that is 65 years and older

The 'Date' column keeps track of COVID-19 cases,deaths,vaccinations,tests from December 2019 till March 25 2021. These columns (Cases,deaths etc) along side the columns in 9,10,11 will further help in our anlysis as described in our business case. We wilL use the 'population' column to calculate the % cases, tests,vaccinations and deaths. 




Thirdly,

This dataset was downloaded from [kagle](https://www.kaggle.com/paultimothymooney/latitude-and-longitude-for-every-country-and-state) website. This dataset has "latitude" and "longitude" which we will need later on to plot a map chart. We will select the following columns;

1.   Country code - iso-code
2.   Latitude
3.   Longitude

The country code will help us join the table with the already existing tables. While the latitude and longitude will be used to represent each individual country on the map.



# DATA PREPARATION

In [ ]:
# we are going to import a data set from our github repository
un_data = pd.read_csv('https://raw.githubusercontent.com/ogtaylor2008837/Femi-s-Bucket/main/country_profile_variables.csv')
un_data

country  ... Net Official Development Assist. received (% of GNI)
0                  Afghanistan  ...                                                -99  
1                      Albania  ...                                                -99  
2                      Algeria  ...                                                -99  
3               American Samoa  ...                                                -99  
4                      Andorra  ...                                                -99  
..                         ...  ...                                                ...  
224  Wallis and Futuna Islands  ...                                                -99  
225             Western Sahara  ...                                                -99  
226                      Yemen  ...                                                -99  
227                     Zambia  ...                                                -99  
228                   Zimbabwe  ...                                                -99  

[229 rows x 50 columns]

After exploring the data set, we wil now begin the data preparation process for this dataset by keeping only the columns relevant for our analysis.


In [ ]:
# this codes keeps only the required columns
un_dataset = un_data[['country', 
                      'GDP: Gross domestic product (million current US$)', 
                      'GDP per capita (current US$)', 'Life expectancy at birth (females/males, years)', 
                      'Health: Total expenditure (% of GDP)']]
un_dataset

country  ...  Health: Total expenditure (% of GDP)
0                  Afghanistan  ...                                   8.2
1                      Albania  ...                                   5.9
2                      Algeria  ...                                   7.2
3               American Samoa  ...                                 -99.0
4                      Andorra  ...                                   8.1
..                         ...  ...                                   ...
224  Wallis and Futuna Islands  ...                                 -99.0
225             Western Sahara  ...                                 -99.0
226                      Yemen  ...                                   5.6
227                     Zambia  ...                                   5.0
228                   Zimbabwe  ...                                   6.0

[229 rows x 5 columns]

Now we continue the data preparation process by renaming every column with a more appropriate name for each unclear column. We will also split the 'Life expectancy at birth (females /males, years)' column into individual male and female columns.

In [ ]:
# we rename the columns appropriately
un_dataset = un_dataset.rename(columns={'country' : 'Country', 
                                        'GDP per capita (current US$)' : 'GDP Per Capita (USD)', 'GDP: Gross domestic product (million current US$)' : 'GDP (million USD)', 
                                        'Health: Total expenditure (% of GDP)' : 'Total Health Expenditure (% of GDP)',})

# we split the 'Life expectancy' column into two
un_dataset[['Female Life Expectancy At Birth (YRS)', 'Male Life Expectancy At Birth (YRS)']] = un_dataset['Life expectancy at birth (females/males, years)'].str.split('/', expand = True)
un_dataset = un_dataset.drop(columns=['Life expectancy at birth (females/males, years)'])

un_dataset

Country  ...  Male Life Expectancy At Birth (YRS)
0                  Afghanistan  ...                                 61.0
1                      Albania  ...                                 75.6
2                      Algeria  ...                                 74.1
3               American Samoa  ...                                 71.1
4                      Andorra  ...                                 None
..                         ...  ...                                  ...
224  Wallis and Futuna Islands  ...                                 72.8
225             Western Sahara  ...                                 66.9
226                      Yemen  ...                                 62.8
227                     Zambia  ...                                 57.5
228                   Zimbabwe  ...                                 56.1

[229 rows x 6 columns]

After futher exploration of the data, it was clear we had to remove the few rows with null values as well as rows with a Total Health Expenditure less than zero (as this shows the values are inaccurate as there can be no health expenditure less than zero or in the negative). We have also changed each columns to their appropriate variable types while also replacing an invalid number "-99.0  to Nan in some columns.

In [ ]:
# this removes rows with null values and values less than or equal to 0 in 'Total Health Expenditure'
un_dataset = un_dataset[un_dataset['Male Life Expectancy At Birth (YRS)'].notna()]
un_dataset = un_dataset[un_dataset['Female Life Expectancy At Birth (YRS)'].notna()]
un_dataset = un_dataset[un_dataset['Total Health Expenditure (% of GDP)'] > 0]

# Here we convert the columns types appropriately
un_dataset['Male Life Expectancy At Birth (YRS)'] = pd.to_numeric(un_dataset['Male Life Expectancy At Birth (YRS)'], errors='coerce')
un_dataset['Total Health Expenditure (% of GDP)'] = pd.to_numeric(un_dataset['Total Health Expenditure (% of GDP)'], errors='coerce')
un_dataset['Female Life Expectancy At Birth (YRS)'] = pd.to_numeric(un_dataset['Female Life Expectancy At Birth (YRS)'], errors='coerce')

un_dataset

Country  ...  Male Life Expectancy At Birth (YRS)
0                           Afghanistan  ...                                 61.0
1                               Albania  ...                                 75.6
2                               Algeria  ...                                 74.1
5                                Angola  ...                                 57.4
7                   Antigua and Barbuda  ...                                 73.3
..                                  ...  ...                                  ...
222  Venezuela (Bolivarian Republic of)  ...                                 69.9
223                            Viet Nam  ...                                 70.7
226                               Yemen  ...                                 62.8
227                              Zambia  ...                                 57.5
228                            Zimbabwe  ...                                 56.1

[187 rows x 6 columns]

After preparing the 'un_dataset', we can see the index column as been set as the 'Location Column' and we now have a final table of 187 columns and 7 rows below.


In [ ]:
# sets index
un_dataset = un_dataset.set_index('Country')
un_dataset

GDP (million USD)  ...  Male Life Expectancy At Birth (YRS)
Country                                                ...                                     
Afghanistan                                     20270  ...                                 61.0
Albania                                         11541  ...                                 75.6
Algeria                                        164779  ...                                 74.1
Angola                                         117955  ...                                 57.4
Antigua and Barbuda                              1356  ...                                 73.3
...                                               ...  ...                                  ...
Venezuela (Bolivarian Republic of)             344331  ...                                 69.9
Viet Nam                                       193241  ...                                 70.7
Yemen                                           29688  ...                                 62.8
Zambia                                          21255  ...                                 57.5
Zimbabwe                                        13893  ...                                 56.1

[187 rows x 5 columns]

We will now import another dataset for our analysis and save it in a variable named "who_covid". 

In [ ]:
#This imports the dataset from our github repository
#who_covid = pd.read_csv('https://raw.githubusercontent.com/ogtaylor2008837/Femi-s-Bucket/main/owid-covid-data.csv', parse_dates=['date'])
#who_covid

who_covid = pd.read_csv('https://raw.githubusercontent.com/ogtaylor2008837/Femi-s-Bucket/main/owid-covid-data%20(2)%20updated%20march25.csv', parse_dates=['date'])
who_covid

iso_code continent  ... life_expectancy human_development_index
0          AFG      Asia  ...           64.83                   0.511
1          AFG      Asia  ...           64.83                   0.511
2          AFG      Asia  ...           64.83                   0.511
3          AFG      Asia  ...           64.83                   0.511
4          AFG      Asia  ...           64.83                   0.511
...        ...       ...  ...             ...                     ...
77277      ZWE    Africa  ...           61.49                   0.571
77278      ZWE    Africa  ...           61.49                   0.571
77279      ZWE    Africa  ...           61.49                   0.571
77280      ZWE    Africa  ...           61.49                   0.571
77281      ZWE    Africa  ...           61.49                   0.571

[77282 rows x 59 columns]

After exploring the dataset, we noticed it contains values for 'World', 'Africa, 'Europe' etc in the 'location' column which is supposed to be filled with just countries. So we will drop every row whose value is not a country by using the "pycountry" library we imported earlier to create a 2 letter country ISOCODE which only generates if the name of the element in the "Location" column and row is a valid country.

In [ ]:
# we define a function to create country iso code
def do_fuzzy_search(country):
    try:
        result = pycountry.countries.search_fuzzy(country)
    except Exception:
        return np.nan
    else:
        return result[0].alpha_2

iso_map = {country: do_fuzzy_search(country) for country in who_covid["location"].unique()}
who_covid["ISO code 2"] = who_covid["location"].map(iso_map)
who_covid = who_covid[who_covid['ISO code 2'].notna()]
who_covid

iso_code continent  ... human_development_index ISO code 2
0          AFG      Asia  ...                   0.511         AF
1          AFG      Asia  ...                   0.511         AF
2          AFG      Asia  ...                   0.511         AF
3          AFG      Asia  ...                   0.511         AF
4          AFG      Asia  ...                   0.511         AF
...        ...       ...  ...                     ...        ...
77277      ZWE    Africa  ...                   0.571         ZW
77278      ZWE    Africa  ...                   0.571         ZW
77279      ZWE    Africa  ...                   0.571         ZW
77280      ZWE    Africa  ...                   0.571         ZW
77281      ZWE    Africa  ...                   0.571         ZW

[72228 rows x 60 columns]

Now, we are going to pick and display only columns relevant for our analysis and also rename the columns appropriately. We will also replace the name of the country "United State" to its full name to enable us join the table easily with other tables moving forward.

In [ ]:
# this selects the relevant columns
who_covid = who_covid[['date', 'iso_code', 'ISO code 2', 'continent', 'location', 'new_cases', 'new_deaths',
                       'new_tests', 'new_vaccinations','hospital_beds_per_thousand', 'population', 'aged_65_older']]

# this renames the columns
who_covid_1 = who_covid.rename(columns={'iso_code' : 'ISO_Code', 'continent' : 'Continent', 
                                      'location' : 'Country', 'date' : 'Date', 
                                      'new_cases' : 'Cases', 'new_deaths': 'Deaths', 
                                      'new_tests' : 'Tests', 'new_vaccinations' : 'Vaccinations', 
                                      'median_age' : 'Median Age', 'aged_65_older' : 'Aged 65 & Older', 
                                      'extreme_poverty' : 'Extreme Poverty', 'cardiovasc_death_race' : 'Cardiovasc Death Race', 
                                      'human_development_index' : 'Human Dev Index', 'population' : 'Population (2020)', 
                                      'population_density' : 'Population Density (2020)', 'hospital_beds_per_thousand' : 'Hospital Beds Per Thousand',
                                      'handwashing_facilities' : 'Handwashing Facilities'})

# replaces a country name
who_covid_1.replace(to_replace=['United States'], value= ['United States of America'], inplace=True)



#returns negative values to positive values using absolute values
who_covid_1
who_covid_1['Cases'] = abs(who_covid_1['Cases'])
who_covid_1



Date ISO_Code  ... Population (2020) Aged 65 & Older
0     2020-02-24      AFG  ...        38928341.0           2.581
1     2020-02-25      AFG  ...        38928341.0           2.581
2     2020-02-26      AFG  ...        38928341.0           2.581
3     2020-02-27      AFG  ...        38928341.0           2.581
4     2020-02-28      AFG  ...        38928341.0           2.581
...          ...      ...  ...               ...             ...
77277 2021-03-21      ZWE  ...        14862927.0           2.822
77278 2021-03-22      ZWE  ...        14862927.0           2.822
77279 2021-03-23      ZWE  ...        14862927.0           2.822
77280 2021-03-24      ZWE  ...        14862927.0           2.822
77281 2021-03-25      ZWE  ...        14862927.0           2.822

[72228 rows x 12 columns]

we are now going to import a new dataset gotten from [link text](https://www.kaggle.com/paultimothymooney/latitude-and-longitude-for-every-country-and-state). This dataset has "latitude" and "longitude" which we will need later on to plot a map chart



In [ ]:
# imports the dataset
lat_lon = pd.read_csv('https://raw.githubusercontent.com/ogtaylor2008837/Femi-s-Bucket/main/world_country_and_usa_states_latitude_and_longitude_values.csv')
lat_lon = lat_lon[["country_code", "latitude", "longitude", "country"]]

lat_lon

country_code   latitude  longitude               country
0             AD  42.546245   1.601554               Andorra
1             AE  23.424076  53.847818  United Arab Emirates
2             AF  33.939110  67.709953           Afghanistan
3             AG  17.060816 -61.796428   Antigua and Barbuda
4             AI  18.220554 -63.068615              Anguilla
..           ...        ...        ...                   ...
240           YE  15.552727  48.516388                 Yemen
241           YT -12.827500  45.166244               Mayotte
242           ZA -30.559482  22.937506          South Africa
243           ZM -13.133897  27.849332                Zambia
244           ZW -19.015438  29.154857              Zimbabwe

[245 rows x 4 columns]

We will now set the 'Country' column of the 'who_covid table as index so as to enable us join the "who_covid dataframe with the initially imported "un_data" dataframe based on their common columns which are the countries. After joining the two tables, we will then reset the index column. We will also merge the joined table with the above lon_lat table with the ISO code 2 as the common columns.

In [ ]:
# sets index
who_covid_1 = who_covid_1.set_index('Country')

# this joins the 2 tables & resets the index
full_data = who_covid_1.join(un_dataset, how='inner', on='Country')
full_data = full_data.reset_index()
full_data = full_data.merge(lat_lon, how='inner', left_on='ISO code 2', right_on='country_code')
full_data












Country       Date ISO_Code  ...   latitude  longitude      country
0      Afghanistan 2020-02-24      AFG  ...  33.939110  67.709953  Afghanistan
1      Afghanistan 2020-02-25      AFG  ...  33.939110  67.709953  Afghanistan
2      Afghanistan 2020-02-26      AFG  ...  33.939110  67.709953  Afghanistan
3      Afghanistan 2020-02-27      AFG  ...  33.939110  67.709953  Afghanistan
4      Afghanistan 2020-02-28      AFG  ...  33.939110  67.709953  Afghanistan
...            ...        ...      ...  ...        ...        ...          ...
61778     Zimbabwe 2021-03-21      ZWE  ... -19.015438  29.154857     Zimbabwe
61779     Zimbabwe 2021-03-22      ZWE  ... -19.015438  29.154857     Zimbabwe
61780     Zimbabwe 2021-03-23      ZWE  ... -19.015438  29.154857     Zimbabwe
61781     Zimbabwe 2021-03-24      ZWE  ... -19.015438  29.154857     Zimbabwe
61782     Zimbabwe 2021-03-25      ZWE  ... -19.015438  29.154857     Zimbabwe

[61783 rows x 21 columns]

Now we will resample the dataframe so that the date record for each country will now be monthly instead of daily while also creating new columns for the % per population for covid cases, deaths, tests and vaccinations. This will be achieved by dividing the cases by population and then multiplying by 100.

In [ ]:
# this resamples to monthly date while grouping by location
full_data = pd.DataFrame(full_data.set_index('Date').groupby('Country').resample('M').agg({'ISO_Code':'last','Continent':'last', 'Cases':'sum', 'Deaths':'sum', 
                                                                                            'Tests':'sum', 'Vaccinations':'sum', 'Aged 65 & Older':'last', 'Population (2020)':'last', 
                                                                                            'GDP (million USD)':'last', 'GDP Per Capita (USD)':'last', 
                                                                                           'Total Health Expenditure (% of GDP)':'last', 
                                                                                            'Female Life Expectancy At Birth (YRS)':'last', 'Male Life Expectancy At Birth (YRS)':'last', 
                                                                                           'latitude':'last', 
                                                                                            'longitude':'last', 'Hospital Beds Per Thousand':'last'}))
# this runs the calculations for the new columns
full_data['% Cases per Population'] = (full_data['Cases'] / full_data['Population (2020)'] * 100).round(2)
full_data['% Deaths per Population'] = (full_data['Deaths'] / full_data['Population (2020)'] * 100).round(2)
full_data['% Vaccinations per Population'] = (full_data['Vaccinations'] / full_data['Population (2020)'] * 100).round(2)
full_data['% Tests per Population'] = (full_data['Tests'] / full_data['Population (2020)'] * 100).round(2)

full_data



ISO_Code  ... % Tests per Population
Country     Date                 ...                       
Afghanistan 2020-02-29      AFG  ...                   0.00
            2020-03-31      AFG  ...                   0.00
            2020-04-30      AFG  ...                   0.00
            2020-05-31      AFG  ...                   0.00
            2020-06-30      AFG  ...                   0.00
...                         ...  ...                    ...
Zimbabwe    2020-11-30      ZWE  ...                   0.19
            2020-12-31      ZWE  ...                   0.32
            2021-01-31      ZWE  ...                   0.69
            2021-02-28      ZWE  ...                   0.42
            2021-03-31      ZWE  ...                   0.29

[2148 rows x 20 columns]

Below, we will further resample the monthly records to a 16 months record which is the entire period of the dates column in our dataframe, so as to get the total cases, deaths, tests, vaccinations as well as their % per Population values for each country.  

In [ ]:
# resets the index before resampling
full_data = full_data.reset_index()
period_16months = pd.DataFrame(full_data.set_index('Date').groupby('Country').resample('16M').agg({'ISO_Code':'last','Continent':'last', 'Cases':'sum', 'Deaths':'sum', 
                                                                                            'Tests':'sum', 'Vaccinations':'sum', 'Aged 65 & Older':'last', 'Population (2020)':'last', 
                                                                                            'GDP (million USD)':'last', 'GDP Per Capita (USD)':'last', 
                                                                                           'Total Health Expenditure (% of GDP)':'last', 
                                                                                            'Female Life Expectancy At Birth (YRS)':'last', 'Male Life Expectancy At Birth (YRS)':'last', 
                                                                                           'latitude':'last', 
                                                                                            'longitude':'last', 'Hospital Beds Per Thousand':'last', '% Cases per Population':'sum', 
                                                                                            '% Deaths per Population':'sum', '% Tests per Population':'sum',
                                                                                            '% Vaccinations per Population':'sum'}))

period_16months




ISO_Code  ... % Vaccinations per Population
Country     Date                 ...                              
Afghanistan 2020-02-29      AFG  ...                          0.00
            2021-06-30      AFG  ...                          0.00
Albania     2020-02-29      ALB  ...                          0.00
            2021-06-30      ALB  ...                          0.49
Algeria     2020-02-29      DZA  ...                          0.00
...                         ...  ...                           ...
Yemen       2021-08-31      YEM  ...                          0.00
Zambia      2020-03-31      ZMB  ...                          0.00
            2021-07-31      ZMB  ...                          0.00
Zimbabwe    2020-03-31      ZWE  ...                          0.00
            2021-07-31      ZWE  ...                          0.36

[322 rows x 20 columns]

From the output above, it is clear why we chose a 16 months period which is to ensure that all values for every country are properly resampled because some countries did not start having values until March 2020 eg. Zimbabwe.
Now to display single final values for each countries we will select the last dates being the end of the 16 months period for each country.

In [ ]:

period_16months = period_16months.reset_index()
period_16months= period_16months.set_index('Date')
period_16months
complete = period_16months[(period_16months.index == '2021-06-30') + (period_16months.index == '2021-07-31') + (period_16months.index == '2021-05-31')]





complete
#print(complete.to_string())



Country  ... % Vaccinations per Population
Date                                  ...                              
2021-06-30               Afghanistan  ...                          0.00
2021-06-30                   Albania  ...                          0.49
2021-06-30                   Algeria  ...                          0.00
2021-07-31                    Angola  ...                          0.00
2021-07-31       Antigua and Barbuda  ...                          0.00
...                              ...  ...                           ...
2021-05-31  United States of America  ...                         36.47
2021-07-31                   Uruguay  ...                         13.03
2021-07-31                Uzbekistan  ...                          0.00
2021-07-31                    Zambia  ...                          0.00
2021-07-31                  Zimbabwe  ...                          0.36

[152 rows x 21 columns]

We will now select the top 30 countries with the highest GDP.

In [ ]:
complete_TOP = complete[complete['GDP (million USD)'] >=  301308]    #top30
complete_TOP.sort_values(by='GDP (million USD)')


Country  ... % Vaccinations per Population
Date                                  ...                              
2021-06-30                   Denmark  ...                         17.18
2021-06-30              South Africa  ...                          0.35
2021-06-30                     Egypt  ...                          0.00
2021-05-31      United Arab Emirates  ...                         69.81
2021-06-30                   Austria  ...                         15.45
2021-06-30                    Norway  ...                         15.21
2021-05-31                  Thailand  ...                          0.01
2021-06-30                   Belgium  ...                         13.47
2021-07-31                    Poland  ...                          9.80
2021-06-30                   Nigeria  ...                          0.00
2021-06-30                    Sweden  ...                          0.90
2021-05-31                 Argentina  ...                          6.40
2021-07-31              Saudi Arabia  ...                          8.41
2021-06-30               Switzerland  ...                         12.14
2021-07-31                    Turkey  ...                         16.87
2021-06-30               Netherlands  ...                          0.00
2021-07-31                 Indonesia  ...                          2.68
2021-05-31                    Mexico  ...                          4.80
2021-06-30                     Spain  ...                         10.12
2021-05-31                 Australia  ...                          1.61
2021-05-31                    Canada  ...                         12.15
2021-06-30                    Brazil  ...                          7.33
2021-05-31                     Italy  ...                         14.48
2021-05-31                     India  ...                          3.99
2021-05-31                    France  ...                         12.32
2021-05-31            United Kingdom  ...                         42.84
2021-05-31                   Germany  ...                         14.00
2021-05-31                     Japan  ...                          0.47
2021-05-31                     China  ...                          0.76
2021-05-31  United States of America  ...                         36.47

[30 rows x 21 columns]

Because we will need to display the covid cases trend for the top 10 GDP Countries in a Bar Chart Race, we will select the top 10 GDP countries and also pivot the table by 'Date and 'Country' with 'Cases' as the values. 

In [ ]:
# Extracts GDP greater than 1552808
full_data_TOP_10 = full_data[full_data['GDP (million USD)'] >= 1552808]

# this pivots the table
fulltop_10_pivot = full_data_TOP_10.pivot("Date", columns="Country", values=["Cases"])     
fulltop_10_pivot

Cases            ...                                        
Country        Brazil    Canada  ... United Kingdom United States of America
Date                             ...                                        
2020-01-31        NaN       4.0  ...            2.0                      7.0
2020-02-29        2.0      16.0  ...           59.0                     17.0
2020-03-31     5715.0    8507.0  ...        38754.0                 192276.0
2020-04-30    81470.0   45930.0  ...       139956.0                 888719.0
2020-05-31   427662.0   38022.0  ...        78768.0                 717698.0
2020-06-30   887192.0   13618.0  ...        27677.0                 843456.0
2020-07-31  1260444.0   12184.0  ...        19577.0                1925246.0
2020-08-31  1245787.0   12637.0  ...        33290.0                1459476.0
2020-09-30   902663.0   30189.0  ...       117765.0                1208476.0
2020-10-31   724670.0   76206.0  ...       558947.0                1930167.0
2020-11-30   800273.0  144244.0  ...       618941.0                4504771.0
2020-12-31  1340095.0  202852.0  ...       862499.0                6428490.0
2021-01-31  1528758.0  199286.0  ...      1331952.0                6147005.0
2021-02-28  1346528.0   87999.0  ...       360640.0                2400568.0
2021-03-31  1768910.0   84848.0  ...       144095.0                1432909.0

[15 rows x 10 columns]

# DATA VISUALIZATION

Now that we have properly selected and prepared our datasets, to support and shed more light on our business case, we will now visualize our datasets by plotting necessary and relevant graphs/charts.

Firstly we will take a look at the trend of Covid-19 cases between January 2020 and March 2021 for the top 10 countries with the highest GDP by plotting a Bar Chart Race.

In [ ]:
# this plots  the Bar Chart Race
bcr.bar_chart_race(fulltop_10_pivot, title = "Covid-19 Cases By Top GDP Countries", n_bars=10, 
                   orientation='h', fixed_order=False, cmap=None, fixed_max=True, steps_per_period=20, 
                   period_length=2350, interpolate_period=False, 
                   period_label={'x': .98, 'y': .3, 'ha': 'right', 'va': 'center'}, 
                   period_summary_func=lambda v, r: {'x': .98, 'y': .2, 
                                          's': f'Total Cases: {v.sum():,.0f}', 
                                          'ha': 'right', 'size': 11}, 
                   bar_size=.95, shared_fontdict=None, scale='linear', 
                   fig=None, writer=None, bar_kwargs={'alpha': .7}, 
                   filter_column_colors=False)

From the Bar Chart race above, we can see that initially by early January 2020, China had the highest cases of just roughly over 9000 with India having the least case of 1 while the United States was just 3rd in the race with just a low case of 7. Towards the summer of 2020, Brazil was at number one slightly ahead of the United States with both countries having over 800,000 cases. It can also be seen that the cases have dropped drastically in China, while China was now at the bottom of the pack,cases in India had increased significantly to almost 400,000 cases and sitting comfortably at number 3. By the end of 2020, the United States was topping the charts with the highest cases of over 6,000,000 while the cases un the Uk was rising again and spotted on number 3, China ended the year still on low cases at bottom. By the end of February we could see that the cases for every country was now dropping. It is interesting to see the increase and decrease of covid cases for these countries from the begining of the pandemic till recently.



Now with the help of a scatter plot, we will visualize and analyse the Covid-19 Vaccinations by the Top 30 GDP countries in relation to the Covid cases between December 2020 and the 9th of March 2020.

In [ ]:
top_30 = full_data[full_data['GDP (million USD)'] >= 301308]
top_30 = top_30.reset_index()

vacc = top_30[(top_30['Date'] >= "2020-12-01")]    
vacc['Date'] = vacc['Date'].dt.month_name()
                               
plot = px.scatter(vacc, 
           x="% Vaccinations per Population", 
           y="% Cases per Population", 
           size="GDP (million USD)",
           animation_frame = "Date",
           animation_group = "Country",
           log_x = True, 
           color="Continent",
           size_max=50,    
           title='% Cases per population vs % Vaccinations per Population by GDP (million USD)', 
           hover_name="Country",
           labels={"Total Health Expenditure (% of GDP)":"Health Expenditure (%)"})
plot.layout.updatemenus[0].buttons[0].args[1]["frame"]["duration"] = 3000
plot.show()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


The scatterplot is designed to show the rate of cases per population against the rate of vaccination per population. The plot also shows impact of the GDP on Vaccination. In December, we can see a that most European countries and North American Countries have started the Vaccination, with Denmark, taking the lead. Between January and February, we can see that Saudi Arabia, who was static in Vaccination, had an increased case of 0.01%. Between February and March, Japan who had increased from Vaccinating 0.01% of its population, to 0.46% of its population had a decline of the rate of cases from 0.03% to 0.02%. Also, within February and March, South Africa with an increased vaccination rate, from 0.12% to 0.23% has seen a decline in cases from a rate of 0.1% to 0.05%. However, a conclusion cannot be drawn on this alone. But from the plot, we can see that the GDP has little or effect on the progress of the vaccination.

Now we will plot a map plot relating the % of deaths per population of the top 30 GDP countries by the hospital beds available per thousand people of the population. We will see the impact of hospital bed availability on covid deaths per population.

In [ ]:
view = px.scatter_mapbox(complete_TOP.reset_index(),
                        lat="latitude",
                        lon="longitude", 
                        color="Hospital Beds Per Thousand", 
                        hover_name="Country",     
                        zoom=1,
                        size='% Deaths per Population',
                        size_max=20,
                        title='Hospital Beds per thousand vs Deaths per Population by Country',
                        )
#view.update_layout(margin={"r":0,"t":0,"l":0,"b":0})
view.show()

From the map plot above, we can see that Germany has the highest hospital beds per thousand but still has roughly the same % deaths per population with a country like Neatherlands which has a much lower hospital bed per thousand. Canada has a low %dpp but also a low hospital beds per thousand just like South Africa. Denmark and Turkey also have both low %dpp as well as Hospital beds per thousand. Argentina with a slightly higher hospital bed per thousand than the United States has a lower % dpp than the united states. It is clear that the amount of hospital beds per thousand in a country will not determine if the % covid deaths per population will be high or low.

Now we will plot a bar chart that informs us about the reationship between the % of the population that died of Covid in the top 30 GDP countries and the % of the GDP these countries spent on Health. Some countries in the top 30 GDP will not make it to this chart because the % of covid deaths per population for these countries are highly insignificant as they are as good as zero. This tells us very few people died in these countrieswhen compared to the population 

In [ ]:
complete_TOP = complete_TOP[complete_TOP['% Deaths per Population'] > 0]    #this helps display only countries with %DpP greater than 0
complete_TOP = complete_TOP.reset_index()
 
barplot = px.bar(complete_TOP, 
           x="Country",
           y="% Deaths per Population",
           color="Total Health Expenditure (% of GDP)", 
           color_continuous_scale='aggrnyl',
           title='% Deaths per Population  vs  Country by Total Health Expenditure (% of GDP)', 
           hover_name="Country")
barplot.update_xaxes(categoryorder='total ascending')

From the bar chart above, we can see that when compared to the other countries, the United States of America spent the most amount of over 16% from their GDP on health while still having the 3rd highest % death per population. While the United Kingdom which is the country with the highest % death per population with slightly over 9% of their GDP spent on Health, Belgium on the other hand with similar %dpp with the United Kingdom had a % health expenditure of 10.6%.
Countries like Swedeen and Switzerland have rougly around 11.6% GDP spend on health which is lower than that of the United States but they still have lower % deaths compared to the United States who spent way higher than them on health, however, they have lower deaths when compared to countries ike Spain and Italy who spend lower than them on health
Austria and Germany who have similar health spend and %dpp have higher %deaths than South Africa and Canada who spent lower on health. It is not very distinct that % Total Expenditure of the GDP have a strong imapct in the rise of Covid Deaths per Population.

Below, we will plot a treemap that shows how the  top 50 countries with the highest GDP's % Deaths per population perfomed in relation with Male life Expectancy. Also note that not all countries will be displayed as the values for some are as good as zero. As a result of this, only values greater than Zero will be displayed.                               

In [ ]:
complete = complete[complete['% Deaths per Population'] > 0]
top_50_count = complete[complete['GDP (million USD)'] >= 164234]
top_50_count["world"] = "world"


px.treemap(top_50_count, path=['world', 'Continent', 'Country'], values='% Deaths per Population',
                  color='Male Life Expectancy At Birth (YRS)', hover_data=['ISO_Code'],
                  color_continuous_scale='RdBu', title='Treemap of Country vs % Deaths per Population by Male Life Expectancy At Birth (YRS)',
                  color_continuous_midpoint=np.average(top_50_count['Male Life Expectancy At Birth (YRS)'], weights=top_50_count['% Deaths per Population']))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



From the above Treemap, we can see that European countries like Czechia, Belgium, United Kingdom and Italy with similar and high Male Life Expectancy have a high % death per population averaging 0.18%. However, countries like Netherlands, Austria, Germany, Ireland and Greece with similar Male life Expectancy  have lower % death per population compared to them. South American Countries like Peru, Argentina and Brazil with lower Male Life Expectancy also have similar % death per population with Germany et al. In North America, USA and Mexico have slightly lower Male Life Expectancy than UK, Belgium and Italy, they have roughly the same % death per population. However Isreal and Canada has one of the highest Male Life Expectancy in the world but have a very low % deaths per population. South Africa has a very low Male Life Expectancy  but interestly has a low % death per population. It is clear that the Male Life Expectancy has no impact on the % of the population that lost their lives to the Covid-19 Pandemic.

It is not a myth that older people may find it harder to survive Covid-19. The sunburst chart will be used to analyse the top 30 GDP countries by the % deaths per population against the portion of a country's population that are Aged 65 and older.

In [ ]:
complete_TOP = complete_TOP[complete_TOP['% Deaths per Population'] > 0]      #this helps display only countries with %DpP greater than 0

px.sunburst(complete_TOP, path=['Continent', 'Country'], values='% Cases per Population',   #add title on all plots beacuse of Lina
                  color='Aged 65 & Older', hover_data=['ISO_Code'],
                  color_continuous_scale='rdbu', title='Country vs %Cases per Population by Aged 65 & Older',
                  color_continuous_midpoint=np.average(complete_TOP['Aged 65 & Older'], weights=complete_TOP['% Cases per Population']))

By exploring the Sunburst chart above, Italy compared to other countries has the highest share of its population at age 65 and older at 23.01 but no where near the % Cases per population of countries like Spain, Belgium, Sweden etc which have slighly lower share of thier population at aged 65 and older when compared to Italy. Denmark and Germany who also has a large share of their population slightly behind Italy have a relatively low % cases per population when compared to other European countries in the chart, as well as the United States of America whose share of its population is around 15. Brazil and Turkey share around the same share per population of 65 years and older, however, they have different % cases per population with Brazil on the higher side. United kingdon and Canada have around the same share of their population aged 65 and older, however their is a clear difference in the % cases per population between the two. It is clear the share of the population has no direct impact on the % cases per population.

Finally, the line graph will shed more light on the percentage of lives lost due to covid per population over a period of time among the top 30  GDP countries

In [ ]:
line_top_30 = full_data[full_data['GDP (million USD)'] >= 301308]
line_top_30 = line_top_30.reset_index()

px.line(line_top_30, x="Date", y="% Deaths per Population", 
        color='Country', title='Deaths per Population vs Date by Continent', facet_col='Continent')

It is shown that Belgium, Argentina, United States of America, Turkey and South Africa have the hghest % deaths per population in their respective continents with Belgium and the United kingdom  having the highest % deaths per population at 0.06 and 0.05 respectively. Argentina, US and South Africa all have the same % deaths per population of 0.03.Countries like Nigeria and the UAE have significantly low Covid deaths, reason why the % deaths per population is closer to zero.